lib/spack/spack/container/writers/__init__.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Writers for different kind of recipes and related
+convenience functions.
+"""
+import collections
+import copy
+
+import spack.schema.env
+import spack.tengine as tengine
+import spack.util.spack_yaml as syaml
+
+from spack.container.images import build_info, package_info
+
+#: Caches all the writers that are currently supported
+_writer_factory = {}
+
+
+def writer(name):
+    """Decorator to register a factory for a recipe writer.
+
+    Each factory should take a configuration dictionary and return a
+    properly configured writer that, when called, prints the
+    corresponding recipe.
+    """
+    def _decorator(factory):
+        _writer_factory[name] = factory
+        return factory
+    return _decorator
+
+
+def create(configuration):
+    """Returns a writer that conforms to the configuration passed as input.
+
+    Args:
+        configuration: how to generate the current recipe
+    """
+    name = configuration['spack']['container']['format']
+    return _writer_factory[name](configuration)
+
+
+def recipe(configuration):
+    """Returns a recipe that conforms to the configuration passed as input.
+
+    Args:
+        configuration: how to generate the current recipe
+    """
+    return create(configuration)()
```

Comment:
It might be clearer to call a `.write()` function here instead of having it hidden in the `.__call__` method. It's slick, but I think the alternative is clearer code.

Reply:
Not particularly tied to it, but I thought using `__call__`  avoids redundancy since I was envisioning clients of this API having code like:
```python
writer = spack.container.writer.create('singularity')
recipe = writer()
```
rather than:
```python
writer = spack.container.writer.create('singularity')
recipe = writer.write()
```
Let me know if you still want to change the name after this explanation and I'll do it.